docs/src/diplomacy/adder_tutorial.md

```diff
@@ -0,0 +1,324 @@
+parameters along **outward edges**. Similarly, `uFn` maps **upward**
+parameters along **outward edges** into **upward** parameters along **inward
+edges**.
+
+```scala mdoc
+/** node for [[Adder]] (nexus) */
+class AdderNode(dFn: Seq[DownwardParam] => DownwardParam,
+                uFn: Seq[UpwardParam] => UpwardParam)(implicit valName: ValName)
+  extends NexusNode(AdderNodeImp)(dFn, uFn)
+```
+
+There are several other node types, which can be found in [Nodes.scala](https://github.com/chipsalliance/rocket-chip/blob/master/src/main/scala/diplomacy/Nodes.scala).
+
+### Creating the `LazyModule`s
+
+"Lazy" evaluation refers to deferring the evaluation of an expression until
+it is needed.
+
+Parameter negotiation in Diplomacy is done lazily, after the Diplomacy graph
+is created. Because of this, the concrete hardware that we want to parameterize
+must also be generated lazily. Diplomacy provides the `LazyModule` construct for
+writing lazily-evaluated hardware modules.
+
+Let's use `LazyModule` to define our `Adder`. Notice that the creation of the
+components used to define Diplomacy graph (in this case, the node), is non-lazy.
+The desired hardware for the module must be written inside `LazyModuleImp`.
+
+For this example, we expect two drivers, each driving an addend of equal bit
+widths into the adder, meaning we expect all of our downward flowing parameters
+to be equivalent. We also expect a sole monitor, so all of our upward flowing
+parameters should also be equivalent. Thus, we can require these properties in
+our `AdderNode` and simply pass on downward the first `DownwardParam` we see.
+Following the same logic, we can pass upward the first `UpwardParam` we see.
+
+```scala mdoc
+/** adder DUT (nexus) */
+class Adder(implicit p: Parameters) extends LazyModule {
+  val node = new AdderNode (
+    { case dps: Seq[DownwardParam] =>
+      require(dps.forall(dp => dp.width == dps.head.width))
+      dps.head
+    },
+    { case ups: Seq[UpwardParam] =>
+      require(ups.forall(up => up.width == ups.head.width))
+      ups.head
+    }
+  )
+  lazy val module = new LazyModuleImp(this) {
+    require(node.in.size == 2)
+    node.out.head._1 := node.in.head._1 + node.in.tail.head._1
+  }
+
+  override lazy val desiredName = "Adder"
+}
+```
+
+Our `AdderDriver` randomly generates an addend of `width` bits each cycle, and
+passes it to `numOutputs` sources.
+
+```scala mdoc
+/** driver (source)
+  * drives one random number on multiple outputs */
+class AdderDriver(width: Int, numOutputs: Int = 2)(implicit p: Parameters) extends LazyModule {
+  val nodeSeq = Seq.fill(numOutputs)(new AdderDriverNode(DownwardParam(width)))
+
+  lazy val module = new LazyModuleImp(this) {
+    // generate random addend (notice the use of the negotiated width)
+    val randomAddend = FibonacciLFSR.maxPeriod(nodeSeq.head.out.head._2.width)
+
+    // drive signals
+    nodeSeq.foreach { node => node.out.head._1 := randomAddend }
```

Comment:
```
 nodeSeq.map(._out).foreach { case (outputBundle, _) => outputBundle := randomAddend }
```
would let this code keep working if you ever changed the number of edges connected to each node

Reply:
Just `node.out.foreach` if you make my above suggested change